Guided Transfer: Save host details

jordwest opened the pull request:
**~~Depends on #7430~~ Merged**

This changes triggers the save host details action (which hits the relevant API endpoint) when the Continue button is pressed on the host details form, then redirects to the checkout.

### How to test

1. Go to **My Sites > Settings > Export**
2. Click **Purchase a Guided Transfer**
3. Choose a host
4. Enter some fake details
5. Click **Continue**
6. You should see an error:
![screen shot 2016-09-02 at 12 10 15 pm](https://cloud.githubusercontent.com/assets/416133/18190947/afe62b54-7109-11e6-8508-a24ff0504889.png)
7. If you have a real Bluehost/SiteGround account for testing, enter the details and you should be redirected to the cart containing the Guided Transfer.

Test live: https://calypso.live/?branch=add/guided-transfer/save-host-details

jordwest commented:
This PR is now complete and ready for review, but it includes and depends on the changes in #7430, so I'll wait until that's merged before changing the status here

dllh commented:
I had a quick test of this, and it redirected to the checkout page as expected. I won't dig much more until #7430 lands and the overlapping code is yanked out of this one.

jordwest commented:
#7430 has now landed, I've rebased this PR and marking it ready for review.

jordwest commented:
Rebased again and resolved conflicts introduced in #7270

dllh commented:
I gave this a test, and it works as expected. The code all looks pretty reasonable to me, but I'm not very familiar with some of the things this code uses, so I'd like to get a proper review from a dev who's got fingers in Calypso code every day.

aduth commented on client/my-sites/exporter/export-card/advanced-settings.jsx:
```diff
@@ -8,7 +8,7 @@ import { connect } from 'react-redux';
  */
 import PostTypeOptions from './post-type-options';
-import SpinnerButton from './spinner-button';
+import SpinnerButton from 'components/button/spinner-button';
```
Few issues with this change:

- Why not create a separate `components/spinner-button` ?
- The contained `className` (`export-card__spinner-button`) does not adhere to [`wpcalypso/jsx-classname-namespace`](https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/lib/rules/jsx-classname-namespace.js)

jordwest replied:
Moved this into `components/spinner-button` in 6f617fe, and fixed up the className

aduth commented on client/my-sites/exporter/guided-transfer-card/complete-purchase-notice.jsx:
```diff
@@ -0,0 +1,42 @@
+import page from 'page';
+
+class CompletePurchaseNotice extends Component {
+	redirectToCart = () => {
```
This function doesn't reference `this`, so you shouldn't need the `= () => {` trick

aduth replied:
In fact, this component could easily be expressed as a [stateless function component](https://facebook.github.io/react/docs/reusable-components.html#stateless-functions) without too much trouble.

aduth replied:
Oh whoops, I overlooked `this.props.siteSlug`. _But_, I'd recommend using `connect` to retrieve this, not passing from the parent component if possible, like so:

```js
export default connect( ( state ) => {
	return {
		siteSlug: getSiteSlug( state, getSelectedSiteId( state ) )
	};
} )( localize( CompletePurchaseNotice ) );
```

See also: 
 - [`getSelectedSiteId`](https://github.com/Automattic/wp-calypso/blob/dcf83067030e472fdc2e60355e7ebf4b8ab8528e/client/state/ui/selectors.js#L28-L36)
- [`getSiteSlug`](https://github.com/Automattic/wp-calypso/blob/dcf83067030e472fdc2e60355e7ebf4b8ab8528e/client/state/sites/selectors.js#L177-L195)

Generally I recommend `connect` as close to the component using the data as possible.

jordwest replied:
I've moved `siteSlug` into the `connect` as suggested in cd86b2b

aduth commented on client/my-sites/exporter/guided-transfer-card/complete-purchase-notice.jsx:
```diff
@@ -0,0 +1,42 @@
+				status="is-warning"
+				showDismiss={ false }
+				text={ translate(
+					`It looks like you've started a Guided Transfer.
```
This violates [`wpcalypso/i18n-no-newlines`](https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/docs/rules/i18n-no-newlines.md) (those line breaks are respected in a template string).

jordwest replied:
Fixed in 2581ce1

aduth replied:
>This violates wpcalypso/i18n-no-newlines (those line breaks are respected in a template string).

I see you fixed them up for the changes implemented here, but there's a few other issues in `exporter/notices.jsx` which should be cleaned up too (here or in a separate pull request).

aduth commented on client/my-sites/guided-transfer/guided-transfer.jsx:
```diff
@@ -86,7 +86,11 @@ export default React.createClass( {
-							? <HostCredentialsPage hostSlug={ this.props.hostSlug } hostInfo={ hostInfo } />
+							? <HostCredentialsPage
+								siteId={ this.props.siteId }
+								siteSlug={ this.props.siteSlug }
```
This goes to show why `connect` close to the component is nice, since we can skip passing the prop through three layers of the hierarchy.

jordwest replied:
Moved `siteSlug` in 11cbce3. We could probably do the same for `siteId` in the other components, but I'll leave that for another PR.

aduth commented on client/my-sites/guided-transfer/host-credentials-page/index.jsx:
```diff
@@ -29,16 +40,34 @@ class HostCredentialsPage extends Component {
 
+	redirectToCart = () => {
+		upgradesActions.addItem( cartItems.guidedTransferItem() );
+		page.redirect( `/checkout/${ this.props.siteSlug }` );
```
I think this is probably fine, but as an alternative, you could consider using [`lib/route`'s `addSiteFragment`](https://github.com/Automattic/wp-calypso/blob/dcf83067030e472fdc2e60355e7ebf4b8ab8528e/client/lib/route/path.js#L50-L64).

jordwest replied:
Ah interesting, I didn't know about that function. I've left this as is for now, I find it a little easier to read for simple cases like this.

aduth commented on client/my-sites/guided-transfer/host-credentials-page/index.jsx:
```diff
@@ -29,16 +40,34 @@ class HostCredentialsPage extends Component {
+		};
+
+		this.props.submit( payload )
+			.then( didSubmit => {
```
Don't depend on an action creator's promise resolution, as this breaks [the unidirectional data flow contract](http://redux.js.org/docs/basics/DataFlow.html).

jordwest replied:
In the early stages of this PR, I'd experimented with a `Redirect` component, but it caused weird behaviour when pressing the Back button. For now I've solved this one with ed6c7dd which just checks for updated props to trigger the redirect.

aduth commented on client/state/sites/guided-transfer/actions.js:
```diff
@@ -90,13 +95,28 @@ export function saveHostDetails( siteId, data ) {
+			if ( error.message ) {
+				dispatch( errorNotice( error.message ) );
+			} else {
+				dispatch( errorNotice( translate( `Whoops, there was an error saving your details. Please 
```
Rather than dispatching the error notice here, I created a notices middleware for this purpose:

https://github.com/Automattic/wp-calypso/blob/master/client/state/notices/middleware.js

See #6645

aduth replied:
Also violates `wpcalypso/i18n-no-newlines`. Use simple concatenation instead of template string.

jordwest replied:
Using notices middleware in 	ece7978

jordwest commented:
Thanks for the reviews @aduth! I've made most of your suggested changes and responded inline

aduth commented on client/my-sites/guided-transfer/host-credentials-page/index.jsx:
```diff
@@ -29,16 +44,35 @@ class HostCredentialsPage extends Component {
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.isAwaitingPurchase === true ) {
```
What other truthy values are you expecting? We're usually not so explicit about these comparisons. Instead, this is just as effective:

```js
if ( nextProps.isAwaitingPurchase ) {
```

jordwest replied:
Removed in 97ebf31

aduth commented on client/my-sites/guided-transfer/host-credentials-page/index.jsx:
```diff
@@ -3,13 +3,28 @@
  */
 import Bluehost from './bluehost';
 import SiteGround from './siteground';
+
```
Was this newline intentional? Not sure it's necessary.

jordwest replied:
It was intentional - I'd intended to separate the sub-components from the library components. You're right though, it's probably unnecessary. Removed in eccc2e2ab0d2c91f9f6e4d9ce68d438c05a5d8d0

aduth commented on client/components/spinner-button/README.md:
```diff
@@ -0,0 +1,38 @@
+import SpinnerButton from 'components/spinner-button';
+
+export default function MyComponent() {
+	return (<SpinnerButton
```
Formatting here is a bit odd (in fact, invalid since there's no closing paranthesis). I'd have expected:

```js
return <SpinnerButton />;
```

jordwest replied:
Another copypasta mistake 😱  Fixed in 8032f95b51c1cd841af0768724d420f1c91da8f3

aduth commented on client/components/spinner-button/index.jsx:
```diff
@@ -51,7 +51,8 @@ export default React.createClass( {
 				{ loading &&
```
While not introduced in this pull request, a few thoughts on this component:

- Was there a design reference for this? I might have expected (and in fact think I've seen in some places in Calypso) that the spinner be contained within the button, not adjacent.
- Accepting `children` as the text to be shown in the button would feel more natural than a `text` prop, though I guess with the duality of the `loadingText` an argument could be made that it's preferable as-is
- Should we have a default for `loadingText`, like `Loading…` ?

These aren't necessarily blockers to this pull request.

jordwest replied:
> Was there a design reference for this?

The spinner button was built from the early designs for the exporter where it's also used (p195om-2Bg-p2):
![screen shot 2016-09-01 at 1 37 06 pm](https://cloud.githubusercontent.com/assets/416133/18154587/38b2523a-7049-11e6-9cc3-bd339b511f9c.png)

Though these designs are over a year old now - @rickybanister any thoughts on changing this spinner?

I'll leave this as is for now so we can launch GT, but if there's a design improvement we can make I'll put together a separate PR to modify it.

rickybanister replied:
I talked to @mtias a bit about this. The design is definitely old. I don't think we put the spinner inside the button anywhere (that I can think of). The spinner doesn't add a whole lot in either case. We could simply remove the spinner and leave the disabled button there. We could even make the disabled button's label more descriptive and say something like 'Generating Export File...'

If we got realtime progress updates on the status of the export job we could display a progress bar here—that would be the most helpful. The spinner is kind of infinite and inspecific so if we don't need to use it, let's not.

Does that help or just muddy the water?

aduth commented on client/components/spinner-button/docs/example.jsx:
```diff
@@ -0,0 +1,27 @@
+ */
+import SpinnerButton from '../index';
+
+export default class SpinnerButtonExample extends Component {
```
The DevDocs screen is picky about the `displayName` of its components for generating the single component view. For example, this should but does not load at the following URL:

http://calypso.localhost:3000/devdocs/design/spinner-button

The exported component _must_ have a display name of `SpinnerButton` for this to work as expected.

jordwest replied:
Fixed in 8032f95b51c1cd841af0768724d420f1c91da8f3

aduth commented on client/components/spinner-button/docs/example.jsx:
```diff
@@ -0,0 +1,27 @@
+		return (
+			<div className="design-assets__group">
+				<h2>
+					<a href="/devdocs/blocks/domain-tip">Spinner Button</a>
```
Expected copypasta, but this URL is wrong. See also: https://github.com/Automattic/wp-calypso/pull/7534/files#r76627275

jordwest replied:
Whoops, you caught me 😄  Fixed in 8032f95b51c1cd841af0768724d420f1c91da8f3

aduth commented on client/components/spinner-button/style.scss:
```diff
@@ -0,0 +1,4 @@
+.spinner-button__spinner {
+	float: right;
+	padding: 8px;
+}
```
Files should end with a newline. I recommend installing an [EditorConfig](http://editorconfig.org/) plugin for your editor, for which we [specify a configuration](https://github.com/Automattic/wp-calypso/blob/master/.editorconfig).

jordwest replied:
Nice, installed the plugin and fixed the newline in 8032f95b51c1cd841af0768724d420f1c91da8f3

aduth commented on client/devdocs/design/index.jsx:
```diff
@@ -141,6 +142,7 @@ let DesignAssets = React.createClass( {
 					<SelectDropdown searchKeywords="menu" />
 					<SocialLogos />
 					<Spinner searchKeywords="loading" />
+					<SpinnerButton />
```
As this is related to other spinners which specify a `searchKeywords`, I might also expect the same `searchKeywords="loading"`

jordwest replied:
Added `loading input submit` keywords in 	8032f95

aduth commented on client/my-sites/exporter/notices.jsx:
```diff
@@ -62,26 +63,8 @@ class Notices extends Component {
+			return <CompletePurchaseNotice />;
 		}
 
 		return null;
```
Not related to this pull request necessarily, but a few thoughts:

- Your function could have been indented one less for improved readability by returning early for the `! this.props.isGuidedTransferAwaitingPurchase` case.
- You could remove this `return` and the function would work just as well, because the inferred `return undefined;` can be handled by React
- The logic of this function is simple enough that you may as well inline it into the render function:

```js
const { isGuidedTransferAwaitingPurchase } = this.props;

return (
	{ this.exportNotice() }
	{ isGuidedTransferAwaitingPurchase && <CompletePurchaseNotice /> }
);
```

jordwest replied:
Ah good catch. I've removed the method and inlined in 9a9700d

aduth commented on client/my-sites/exporter/index.jsx:
```diff
@@ -22,7 +21,6 @@ class Exporter extends Component {
 		const {
 			siteId,
-			siteSlug,
 			isGuidedTransferInProgress,
```
Not introduced in this pull request, but:

```
$ ./node_modules/.bin/eslint client/my-sites/exporter/index.jsx 

/Users/andrew/Documents/Code/wp-calypso/client/my-sites/exporter/index.jsx
  24:4  warning  'isGuidedTransferInProgress' is already declared in the upper scope  no-shadow

✖ 1 problem (0 errors, 1 warning)
```

jordwest replied:
Fixed in 084b68f

aduth commented on client/state/notices/middleware.js:
```diff
@@ -76,6 +78,21 @@ export function onPostSaveSuccess( dispatch, action ) {
 	}
 }
 
+export function onGuidedTransferHostDetailsSaveSuccess( dispatch ) {
```
The module also contains `dispatchSuccess` as a convenience for this pattern.

jordwest replied:
Replaced in 6659778

aduth commented on client/state/sites/guided-transfer/actions.js:
```diff
@@ -80,23 +88,28 @@ export function saveHostDetails( siteId, data ) {
-			siteId,
-			error,
-		} );
+			if ( response.host_details_entered === true ) {
```
See: https://github.com/Automattic/wp-calypso/pull/7534/files#r76625909

jordwest replied:
Removed in 97ebf31

aduth commented on client/state/sites/guided-transfer/test/actions.js:
```diff
@@ -109,7 +114,9 @@ describe( 'actions', () => {
 			nock( 'https://public-api.wordpress.com:443' )
 				.post( `/wpcom/v2/sites/${sampleSiteId}/transfer` )
-				.times( 3 )
+				.times( 2 )
```
Is `times` here dependent on the number of tests you have an number of times it's running on the endpoint? That seems cumbersome to maintain. Maybe instead vary the response by the sample site ID used. Here are a few example site IDs: `2916284`, `77203074`, `77203199`

jordwest replied:
Agreed, it was indeed dependent on the number of tests. I've implemented your suggestion in eccc2e2

jordwest commented:
Thanks again for the very detailed code reviews @aduth! I'm made most of your suggested changes and tidied up a few other things I hadn't caught.

aduth commented on client/my-sites/guided-transfer/host-credentials-page/index.jsx:
```diff
@@ -29,16 +43,35 @@ class HostCredentialsPage extends Component {
 		this.setFieldValue( fieldName, e.target.value );
 	}
 
+	redirectToCart = () => {
```
Not a blocker, but since this is only called in `componentWillReceiveProps`, we don't need to preserve the `this` context with the arrow function. Not really a blocker, and perhaps better to be on the safe side.

aduth commented on client/state/notices/middleware.js:
```diff
@@ -76,6 +78,17 @@ export function onPostSaveSuccess( dispatch, action ) {
+	if ( error && error.message ) {
+		dispatch( errorNotice( error.message ) );
+	} else {
+		dispatch( errorNotice( translate( 'Whoops, there was an error saving your details. Please ' +
```
Wonder how such a long error message appears. Might be a bit of an eyesore?

aduth replied:
>Wonder how such a long error message appears. Might be a bit of an eyesore?

To answer my own question, not _too_ terrible:

![image](https://cloud.githubusercontent.com/assets/1779930/18169430/1581c96a-7028-11e6-9ba8-a45043fe94eb.png)

rickybanister replied:
That's a pretty long and not very helpful error. If the error is related directly to the form submission, we can put the notice inside the form—perhaps below the sectionHeader as a block-level element.

Also we should avoid saying things like 'send us a message' if there is no noticeAction that triggers a contact form or chat—the user has no way to actually send a message from here.

Could the message say 'Your hosting details are incorrect. Please try again or [contact support](#)'

The support link could be inline, or if that's the primary action it can be a `noticeAction`

jordwest replied:
Previously the error messages were being read from the server, so I couldn't include links in the text. In 25feeb4 I've added an inline notice which uses the error code and adds a Get Help link:
![screen shot 2016-09-02 at 12 10 15 pm](https://cloud.githubusercontent.com/assets/416133/18190421/41d2a8b6-7106-11e6-88f7-56416c1afa0d.png)

Or, alternatively, how about the error on top?
![screen shot 2016-09-02 at 11 58 58 am](https://cloud.githubusercontent.com/assets/416133/18190490/9b6a29bc-7106-11e6-8e30-1029a66fcc2f.png)

aduth commented on client/state/sites/guided-transfer/actions.js:
```diff
@@ -80,23 +88,28 @@ export function saveHostDetails( siteId, data ) {
-			} );
+		const failure = error => {
+			dispatch( saveHostDetailsFailure( siteId, error ) );
+			return false;
```
You shouldn't need to specify a value, `return;` would be suffice I think.

jordwest replied:
Removed in 900f6c1

aduth commented:
Code looks good. I tested and received an error message after setting fake details. Not sure whether that's expected considering the details were fake. But at least confirmed the spinner button appeared.

rickybanister commented:
When testing this, as the interface loaded I saw a flash of a dark gray/blue notice in place of the Purchase button for a second. Could we do something with a placeholder there instead?

jordwest commented:
@aduth:
> I tested and received an error message after setting fake details. 

The real account validation was committed on the server just a couple days ago, so that's expected (forgot to update this PR's testing instructions)

@rickybanister:

> a flash of a dark gray/blue notice

Ah that would be the unavailability notice appearing while confirming that GT is available. I've just opened #7828 to discuss that

rickybanister commented:
Since this isn't a particularly tall form, I like it 'above' the form card as well.

rickybanister commented:
Thanks for the updates!

jordwest commented:
I've just rebased on master and fixed some minor issues (05c6522 and 1c31812).

@rickybanister: I also switched the error notices to appear above the form in d9e05dd.

@aduth: do you feel this needs further review or would you be comfortable giving it the thumbs up?

aduth commented on client/my-sites/guided-transfer/host-credentials-page/error-notice.jsx:
```diff
@@ -0,0 +1,45 @@
+			translate( "We've already confirmed your details. Please contact support if you need to change them." ),
+	};
+
+	const errorText = get(
```
A case to be made for a `switch` with `default` here would be that you could avoid the `translate` call for the default value that you're always making here, since `translate` is a non-trivial function call. Not worried enough to warrant any change.

aduth commented:
No other comments on the code, and the "invalid details" error showed for me as expected in testing. LGTM 👍

jordwest commented:
Thanks! I swapped out the `get` for a `switch` statement in 83fcc50.
